File: app/src/main/java/com/dumuzeyn/mp3player/UiFactory.java
package com.dumuzeyn.mp3player;

import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

final class UiFactory {
    private final MainActivityCore host;
    private final ButtonFactory buttons;

    UiFactory(MainActivityCore host) {
        this.host = host;
        this.buttons = new ButtonFactory(host);
    }

    LinearLayout row() {
        LinearLayout row = new LinearLayout(host);
        row.setOrientation(LinearLayout.HORIZONTAL);
        row.setGravity(16);
        return row;
    }

    TextView text(String value, int size, boolean bold) {
        TextView text = new OutlinedTextView(host);
        text.setText(value);
        text.setTextColor(host.fg);
        text.setTextSize(size);
        text.setGravity(16);
        text.setTypeface(null, bold ? Typeface.BOLD : Typeface.NORMAL);
        text.setSingleLine(false);
        host.themeController.applyTextOutline(text);
        return text;
    }

    TextView dialogTitle(String value) {
        return dialogTitle(value, 22);
    }

    TextView dialogTitle(String value, int size) {
        TextView title = text(value, size, true);
        title.setGravity(Gravity.START | Gravity.CENTER_VERTICAL);
        title.setMinHeight(host.dp(46));
        title.setPadding(0, host.dp(4), 0, host.dp(10));
        return title;
    }

    LinearLayout.LayoutParams dialogTitleParams() {
        return new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT);
    }

    void makeMarquee(TextView text) {
        text.setSingleLine(true);
        text.setEllipsize(TextUtils.TruncateAt.MARQUEE);
        text.setMarqueeRepeatLimit(-1);
        text.setSelected(true);
        text.setFocusable(true);
        text.setFocusableInTouchMode(true);
    }

    Button button(String label) {
        return buttons.button(label);
    }

    Button icon(String symbol) {
        return buttons.icon(symbol);
    }

    Button shuffleButton() {
        return buttons.shuffleButton();
    }

    void applyPlainIconStyle(Button button, int color) {
        buttons.applyPlainIcon(button, color);
    }

    void applyPlainIconStyle(Button button) {
        buttons.applyPlainIcon(button,
                host.appearanceState.dark ? Color.rgb(230, 226, 236) : host.primaryText);
    }

    GradientDrawable cardBackground() {
        return cardBackground(host.appearanceState.dialogCardOpacity);
    }

    GradientDrawable cardBackground(int opacity) {
        GradientDrawable drawable = new GradientDrawable();
        drawable.setColor(host.cardSurfaceColor(host.card, opacity));
        drawable.setCornerRadius(host.dp(16));
        drawable.setStroke(host.dp(1), host.cardStroke);
        return drawable;
    }

    void applyCardStyle(View view) {
        view.setBackground(cardBackground());
        TextOutlinePolicy.markCardSurface(view, true);
        view.setElevation(host.dp(1));
    }

    void applyCardStyle(View view, int opacity) {
        view.setBackground(cardBackground(opacity));
        TextOutlinePolicy.markCardSurface(view, true);
        view.setElevation(host.dp(1));
    }

    void applyPrimaryButtonStyle(Button button) {
        buttons.applyPrimary(button);
    }

    void applySecondaryButtonStyle(Button button) {
        buttons.applySecondary(button);
    }

    void applySecondaryButtonStyle(Button button, int opacity) {
        buttons.applySecondary(button, opacity);
    }

    void applyPlayerToolStyle(Button button, boolean active) {
        buttons.applyPlayerTool(button, active);
    }

    void applySeekBarColors(SeekBar seekBar) {
        if (Build.VERSION.SDK_INT >= 21) {
            seekBar.setProgressTintList(ColorStateList.valueOf(host.purple));
            seekBar.setThumbTintList(ColorStateList.valueOf(host.yellow));
            seekBar.setProgressBackgroundTintList(ColorStateList.valueOf(host.purpleSoft));
        }
    }

    LinearLayout.LayoutParams square(int size) {
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(host.dp(size), host.dp(size));
        params.setMargins(host.dp(4), host.dp(4), host.dp(4), host.dp(4));
        return params;
    }

    View spaced(View view) {
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(-1, -2);
        params.setMargins(0, host.dp(2), 0, host.dp(2));
        view.setLayoutParams(params);
        return view;
    }

    void setSurface(View view, int color, boolean outlined) {
        setSurface(view, color, outlined, host.appearanceState.dialogCardOpacity);
    }

    void setSurface(View view, int color, boolean outlined, int opacity) {
        int surfaceColor = color == host.card || color == host.panel
                ? host.cardSurfaceColor(color, opacity)
                : color;
        view.setBackground(rounded(surfaceColor, outlined));
        TextOutlinePolicy.markCardSurface(view, true);
    }

    View lineView() {
        View line = new View(host);
        line.setBackgroundColor(host.line);
        return line;
    }

    ImageView coverView() {
        ImageView cover = new RotatingCoverImageView(host);
        cover.setBackgroundColor(Color.TRANSPARENT);
        return cover;
    }

    FrameLayout shade() {
        SwipeDismissFrameLayout shade = new SwipeDismissFrameLayout(host);
        int channel = host.appearanceState.dark ? 0 : 255;
        shade.setBackgroundColor(Color.argb(190, channel, channel, channel));
        Runnable dismiss = () -> {
            if (shade.getParent() != null) {
                host.overlayHost.removeView(shade);
            }
            host.playerUiController.updateMini();
        };
        shade.setDismissAction(dismiss);
        shade.setOnClickListener(view -> dismiss.run());
        return shade;
    }

    LinearLayout panelCard() {
        LinearLayout panel = new LinearLayout(host);
        panel.setOrientation(LinearLayout.VERTICAL);
        panel.setPadding(host.dp(12), host.dp(12), host.dp(12), host.dp(12));
        applyCardStyle(panel);
        panel.setOnClickListener(view -> { });
        return panel;
    }

    private GradientDrawable rounded(int color, boolean outlined) {
        GradientDrawable drawable = new GradientDrawable();
        drawable.setColor(color);
        drawable.setCornerRadius(host.dp(outlined ? 16 : 14));
        drawable.setStroke(outlined ? 1 : 0, outlined ? host.cardStroke : color);
        return drawable;
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/UiFactory.kt
package com.dumuzeyn.mp3player

import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Outline
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.text.TextUtils
import android.view.Gravity
import android.view.View
import android.view.ViewOutlineProvider
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.TextView
import kotlin.math.min

internal class UiFactory(private val host: MainActivityCore) {
    private val buttons = ButtonFactory(host)

    fun row(): LinearLayout = LinearLayout(host).apply {
        orientation = LinearLayout.HORIZONTAL
        gravity = Gravity.CENTER_VERTICAL
    }

    fun text(value: String, size: Int, bold: Boolean): TextView = OutlinedTextView(host).apply {
        text = value
        setTextColor(host.fg)
        textSize = size.toFloat()
        gravity = Gravity.CENTER_VERTICAL
        setTypeface(null, if (bold) Typeface.BOLD else Typeface.NORMAL)
        setSingleLine(false)
        host.themeController.applyTextOutline(this)
    }

    fun dialogTitle(value: String): TextView = dialogTitle(value, 22)

    fun dialogTitle(value: String, size: Int): TextView = text(value, size, true).apply {
        gravity = Gravity.START or Gravity.CENTER_VERTICAL
        minHeight = host.dp(46)
        setPadding(0, host.dp(4), 0, host.dp(10))
    }

    fun dialogTitleParams(): LinearLayout.LayoutParams = LinearLayout.LayoutParams(
        LinearLayout.LayoutParams.MATCH_PARENT,
        LinearLayout.LayoutParams.WRAP_CONTENT,
    )

    fun makeMarquee(text: TextView) {
        text.setSingleLine(true)
        text.ellipsize = TextUtils.TruncateAt.MARQUEE
        text.marqueeRepeatLimit = -1
        text.isSelected = true
        text.isFocusable = true
        text.isFocusableInTouchMode = true
    }

    fun button(label: String): Button = buttons.button(label)

    fun icon(symbol: String): Button = buttons.icon(symbol)

    fun shuffleButton(): Button = buttons.shuffleButton()

    fun applyPlainIconStyle(button: Button, color: Int) {
        buttons.applyPlainIcon(button, color)
    }

    fun applyPlainIconStyle(button: Button) {
        buttons.applyPlainIcon(
            button,
            if (host.appearanceState.dark) Color.rgb(230, 226, 236) else host.primaryText,
        )
    }

    fun cardBackground(): GradientDrawable = cardBackground(host.appearanceState.dialogCardOpacity)

    fun cardBackground(opacity: Int): GradientDrawable = GradientDrawable().apply {
        setColor(host.cardSurfaceColor(host.card, opacity))
        cornerRadius = host.dp(16).toFloat()
        setStroke(host.dp(1), host.cardStroke)
    }

    fun applyCardStyle(view: View) {
        view.background = cardBackground()
        TextOutlinePolicy.markCardSurface(view, true)
        view.elevation = host.dp(1).toFloat()
    }

    fun applyCardStyle(view: View, opacity: Int) {
        view.background = cardBackground(opacity)
        TextOutlinePolicy.markCardSurface(view, true)
        view.elevation = host.dp(1).toFloat()
    }

    fun applyPrimaryButtonStyle(button: Button) {
        buttons.applyPrimary(button)
    }

    fun applySecondaryButtonStyle(button: Button) {
        buttons.applySecondary(button)
    }

    fun applySecondaryButtonStyle(button: Button, opacity: Int) {
        buttons.applySecondary(button, opacity)
    }

    fun applyPlayerToolStyle(button: Button, active: Boolean) {
        buttons.applyPlayerTool(button, active)
    }

    fun applySeekBarColors(seekBar: SeekBar) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            seekBar.progressTintList = ColorStateList.valueOf(host.purple)
            seekBar.thumbTintList = ColorStateList.valueOf(host.yellow)
            seekBar.progressBackgroundTintList = ColorStateList.valueOf(host.purpleSoft)
        }
    }

    fun square(size: Int): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(host.dp(size), host.dp(size)).apply {
            setMargins(host.dp(4), host.dp(4), host.dp(4), host.dp(4))
        }

    fun libraryArtwork(size: Int = 52): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(host.dp(size), host.dp(size)).apply {
            setMargins(host.dp(4), host.dp(2), host.dp(4), host.dp(2))
        }

    fun spaced(view: View): View {
        view.layoutParams = LinearLayout.LayoutParams(-1, -2).apply {
            setMargins(0, host.dp(2), 0, host.dp(2))
        }
        return view
    }

    fun spacedLibraryCard(view: View): View {
        view.layoutParams = LinearLayout.LayoutParams(-1, libraryCardHeight()).apply {
            setMargins(0, host.dp(2), 0, host.dp(2))
        }
        return view
    }

    fun libraryCardHeight(): Int =
        host.resources.getDimensionPixelSize(R.dimen.library_card_height)

    fun setSurface(view: View, color: Int, outlined: Boolean) {
        setSurface(view, color, outlined, host.appearanceState.dialogCardOpacity)
    }

    fun setSurface(view: View, color: Int, outlined: Boolean, opacity: Int) {
        val surfaceColor = if (color == host.card || color == host.panel) {
            host.cardSurfaceColor(color, opacity)
        } else {
            color
        }
        view.background = rounded(surfaceColor, outlined)
        TextOutlinePolicy.markCardSurface(view, true)
    }

    fun lineView(): View = View(host).apply { setBackgroundColor(host.line) }

    fun coverView(): ImageView = RotatingCoverImageView(host).apply {
        setBackgroundColor(Color.TRANSPARENT)
    }

    fun staticCoverView(): ImageView = ImageView(host).apply {
        scaleType = ImageView.ScaleType.CENTER_CROP
        outlineProvider = object : ViewOutlineProvider() {
            override fun getOutline(view: View, outline: Outline) {
                val radius = if (host.appearanceState.circularCovers) {
                    min(view.width, view.height) * 0.5f
                } else {
                    host.dp(8).toFloat()
                }
                outline.setRoundRect(0, 0, view.width, view.height, radius)
            }
        }
        clipToOutline = true
        setBackgroundColor(Color.TRANSPARENT)
    }

    fun shade(): FrameLayout {
        val shade = SwipeDismissFrameLayout(host)
        val channel = if (host.appearanceState.dark) 0 else 255
        shade.setBackgroundColor(Color.argb(190, channel, channel, channel))
        val dismiss = Runnable {
            if (shade.parent != null) host.overlayHost.removeView(shade)
            host.playerUiController.updateMini()
        }
        shade.setDismissAction(dismiss)
        shade.setOnClickListener { dismiss.run() }
        return shade
    }

    fun panelCard(): LinearLayout = LinearLayout(host).apply {
        orientation = LinearLayout.VERTICAL
        setPadding(host.dp(12), host.dp(12), host.dp(12), host.dp(12))
        applyCardStyle(this)
        setOnClickListener { }
    }

    private fun rounded(color: Int, outlined: Boolean): GradientDrawable =
        GradientDrawable().apply {
            setColor(color)
            cornerRadius = host.dp(if (outlined) 16 else 14).toFloat()
            setStroke(if (outlined) host.dp(1) else 0, if (outlined) host.cardStroke else color)
        }
}
